chore: Add pnpm-workspaces example

BeyondEvil opened the pull request:
Fixes #112

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aspect-build/bazel-examples?pullRequest=114) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aspect-build/bazel-examples?pullRequest=114) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aspect-build/bazel-examples?pullRequest=114) it.</sub>

BeyondEvil commented:
I'm wondering if I should add an explicit dependency on one or more of the `packages` (shared) ilbs. 🤔

jbedard commented on pnpm-workspaces/apps/alpha/BUILD.bazel:
I don't think this is used?

jbedard commented on pnpm-workspaces/apps/alpha/tsconfig.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "../../tsconfig.json"
```
Is it worth having these in each folder? I think less config is better for this example..?

BeyondEvil replied:
Probably not, I can have just the one in `alpha` and convert `beta` to pure js.

jbedard replied:
I think we can just have one tsconfig in the root and the libs+apps can all share that?

👍 for converting one of the apps to pure js just to show the absolute minimum in one of them

BeyondEvil replied:
I might need some help solving that. Will that require `export_files`? So that the root tsconfig is available for alpha? Or is defining a `ts_config` in the root sufficient and setting the `tsconfig` field on `ts_project` to `//:tsconfig`? @jbedard

jbedard replied:
Looks like you got it 👍

BeyondEvil commented:
I'm going to add a proper README as well. The lack (entirely or information content) of such is something that annoys me with some of the other examples.

jbedard commented on pnpm-workspaces/apps/alpha/BUILD.bazel:
```diff
@@ -0,0 +1,32 @@
+    name = "tsconfig",
+    src = "tsconfig.json",
+    deps = ["//:tsconfig"],
+    visibility = ["//visibility:public"],
```
Can drop this line?

BeyondEvil commented on pnpm-workspaces/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# compiled output
+dist
```
Probably not, copy-pasta error from one of the other examples.

jbedard commented on pnpm-workspaces/package.json:
Do you normally launch bazel with pnpm (or yarn, or npm, ...)? I think we can leave these out?

BeyondEvil replied:
I don't, but devs do. But I can remove them for the example. 👍

jbedard replied:
Just to keep the example minimal 👍

jbedard commented on pnpm-workspaces/packages/two/BUILD.bazel:
```diff
@@ -0,0 +1,33 @@
+npm_package(
+    name = "two",
+    srcs = [
+        "src/main.js",
```
Do you think it's clearer to do `src/main.js` or `:twotwo`?

I think if the `ts_project` had multiple files this would only put this one file into the package?

BeyondEvil replied:
Good point!

jbedard replied:
I often do something such as `_ts` or maybe `_two_ts`.

jbedard commented:
> I'm going to add a proper README as well. The lack (entirely or information content) of such is something that annoys me with some of the other examples.

I was thinking the same. Something simple to explain the `npm_package` and the pnpm-workspace linking all the packages together?

jbedard commented:
Lets add it to `.github/workflows/ci.yaml` as well.

BeyondEvil commented on pnpm-workspaces/package.json:
version is not, for `name` I'm not sure, as we reference that name in `deps`.

BeyondEvil commented on pnpm-workspaces/apps/alpha/package.json:
Not sure, will try without.

BeyondEvil commented on pnpm-workspaces/packages/two/BUILD.bazel:
```diff
@@ -0,0 +1,21 @@
+    name = "two",
+    srcs = [
+        "package.json",
+        ":two_js",
```
This worked.

jbedard commented on pnpm-workspaces/tsconfig.json:
```diff
@@ -0,0 +1,8 @@
+  "extends": "@tsconfig/node16-strictest/tsconfig.json",
+  "compileOnSave": false,
+  "compilerOptions": {
+    "noImplicitAny": false
```
Is this required?

BeyondEvil replied:
Nope. Will remove.

BeyondEvil replied:
Yes, because I haven't typed anything in the local TS-libs (this allows implicit any).

BeyondEvil replied:
Removed `compileOnSave` tho.

jbedard commented on pnpm-workspaces/apps/beta/BUILD.bazel:
Was there a reason for depending on this in the root?

BeyondEvil replied:
Not sure I understand the question.

`inspirational-quotes` is used by both `alpha` and `beta`, and to avoid having to declare it twice (in each project package.json), it's declared in the root package.json.

jbedard replied:
I see... lets keep it this way for now but this is something to think about 👍

jbedard replied:
Yeah that all makes sense and you are correct.

I think I was just suggesting we put it in the root by default, where currently you're putting it in each individual package by default (until it is used in more then 1 place)? Then we are consistent (always put it in the root), except the odd case where we need a different version.

BeyondEvil replied:
Ok, how about a compromise? 😅 

We do what you're suggesting (put deps in root package.json by default), but include a locally defined package with a different version? That way we have an example that shows that you can override the version locally and what that would look like in the BUILD-file.

jbedard replied:
I think that's the best solution. If we include the different version in this example lets just make sure we explain it in the readme.

BeyondEvil replied:
Done ✅

jbedard commented on pnpm-workspaces/packages/one/BUILD.bazel:
```diff
@@ -0,0 +1,24 @@
+    srcs = [
+        "package.json",
+        ":one_ts",
+        "//:tsconfig",
```
I don't think this has to be in the `npm_package`? It should only be required for compiling the ts?

jbedard commented on pnpm-workspaces/packages/one/BUILD.bazel:
```diff
@@ -0,0 +1,24 @@
+    srcs = ["src/main.ts"],
+    tsconfig = "//:tsconfig",
+    deps = [
+        "//:node_modules/@types/node",
```
Are any node APIs being used that require this?

BeyondEvil replied:
Yes, the use of `console` requires it:

```
packages/one/src/main.ts(2,3): error TS2584: Cannot find name 'console'. Do you need to change your target library? Try changing the 'lib' compiler option to include 'dom'.
```

jbedard replied:
👍

jbedard commented:
Actually one idea which we should maybe show/recommend in this example... we can put all third-party dependencies+versions in the root package.json to avoid repeating the version, then put `"pkg": "workspace:*"` in the individual package.json files to declare the dependency while avoiding re-declaring the version. Basically the same as the first-party npm packages. I think this would be the ideal scenario (one version, but fine grained deps) and what we should show in the example.

We can also merge the example as-is if you don't want to continue with all these fine-grained details. The example is great already and we can always do more later 👍

jbedard commented:
> we can put all third-party dependencies+versions in the root package.json to avoid repeating the version, then put "pkg": "workspace:*"

I'm not sure if that will work actually. You can test it out but np if it doesn't work or you don't want to try it atm.

BeyondEvil commented:
> Actually one idea which we should maybe show/recommend in this example... we can put all third-party dependencies+versions in the root package.json to avoid repeating the version, then put `"pkg": "workspace:*"` in the individual package.json files to declare the dependency while avoiding re-declaring the version. Basically the same as the first-party npm packages. I think this would be the ideal scenario (one version, but fine grained deps) and what we should show in the example.
> 
> We can also merge the example as-is if you don't want to continue with all these fine-grained details. The example is great already and we can always do more later 👍

Afaik, there are no dependencies that are repeated. The shared ones are in the root package.json.

Or am I misunderstanding?

jbedard commented:
> Afaik, there are no dependencies that are repeated. The shared ones are in the root package.json.

There are none repeated in your current example but a lot of projects will have shared dependencies across packages. It's definitely not required in this example, at least not right away.

BeyondEvil commented:
Please review the latest commit(https://github.com/aspect-build/bazel-examples/pull/114/commits/5013a2c397f63f69588bfe3297bcad13d74531d2) diligently @jbedard as I've removed _a lot_. 🙏

jbedard commented:
I think that's looking pretty good. Anything else you're still thinking of doing before we merge?

BeyondEvil commented:
> I think that's looking pretty good. Anything else you're still thinking of doing before we merge?

I'm happy if you are.

The only thing we don't cover (which might not be necessary as I don't think it would introduce anything new) is the case where one of the libraries (one, two, shared) has 1st and/or 3rd-party dependencies.

jbedard commented:
> The only thing we don't cover (which might not be necessary as I don't think it would introduce anything new) is the case where one of the libraries (one, two, shared) has 1st and/or 3rd-party dependencies.

I thought of that as well. I guess it would show indirect dependencies, both first and third-party deps. Would that complicate it more? I want to keep this as simple as possible so it's easy to point people to it as an example for the basic recommended pnpm setup...

BeyondEvil commented:
I could add it and then you can judge if it is too complicated or not?

jbedard commented:
If you don't mind trying then lets take a look 👍 

Just a dependency from one of the `npm_package`s to the shared package might be enough? Could also add one more dependency on an external one 🤷

BeyondEvil commented:
> If you don't mind trying then lets take a look 👍
> 
> Just a dependency from one of the `npm_package`s to the shared package might be enough? Could also add one more dependency on an external one 🤷

Done ✅ 

Lmk what you think!

jbedard commented:
I think that's good, and getting large enough I think we shouldn't add any more. Looks like the build is failing atm though.

BeyondEvil commented:
I don’t understand why. It passes locally (after `bazel clean`).

jbedard commented on pnpm-workspaces/packages/two/BUILD.bazel:
```diff
@@ -0,0 +1,21 @@
+load("@aspect_rules_js//js:defs.bzl", "js_library")
+load("@aspect_rules_js//npm:defs.bzl", "npm_package")
+
+js_library(
```
Since there's no ts can we just put all the src+deps here into `npm_package`?

jbedard replied:
Could maybe do that in other BUILDs as well.

`js_library` really just groups some files together while passing along js + d.ts related metadata. If all we do is then put that into an `npm_package` then I don't think it's necessary.

BeyondEvil replied:
`npm_package` doesn't have a `deps` attributed. Where do I put the deps from `js_library`?

jbedard commented on pnpm-workspaces/packages/two/src/main.js:
What's the `two = ` here?

BeyondEvil replied:
I belive is just naming the function, but I can probably return an `anonymous` function instead.

BeyondEvil commented:
I can reproduce it on my local machine. I hade local `node_modules` that interfered.

With that said, I have no idea why it's not working @jbedard 🤷‍♂️ 

All I can see is none of `two`s dependencies (neither `first` or `cowsay`) are in the `node_modules` under bazel-out.

I tried adding `npm_link_all_packages` to `two`, but it didn't make any difference.

BeyondEvil commented:
I think I fixed the issue.

This helped me: https://github.com/aspect-build/rules_js/blob/main/e2e/npm_link_package/lib/BUILD.bazel#L9-L13

It was not obvious to me. 😅 

Also addressed your other comments @jbedard 👍

jbedard commented:
Anything else you'd like to do before merging?

BeyondEvil commented:
> Anything else you'd like to do before merging?

Nope, happy if you are! ☺️

Lmk, if you want me to squash/fixup commits. 👍

jbedard commented:
Merged! 👍 
(you can squash when you merge with the github PR UI)

Thanks for putting this together and all the back-and-forth with the comments+changes on the PR! 👍

BeyondEvil commented:
No worries!

Happy to contribute! ☺️👍
